### ISSUE TITLE
Migrate commits from T4C CDO repository in model backup job

### ISSUE THREAD

**Issue by ewuerger** [DSD-DBS/capella-dockerimages#189]
Users want to see their original commits in the history of their GitLab repository. Currently the backup job is just doing one commit and all individual commits with their messages are not included in the model repository on GitLab. Having the original commits migrated, enables many features/actions available on GitLab. For e.g. issues on the model repo, or Jira tickets can be referenced and their status can be changed (e.g. closed).

**Comment by dominik003** [comment]
I just took the time to take a look at this issue to see how we can implement it.

First of all, when using the importer script (i.e., retrieving a project from the T4C server and then pushing it to a Git repository), we have already implemented the ability to include commit information in two files.

The first file, the `CommitHistory_[...]_.activitymetadata`, has an xml-like structure in which each commit has its own `activities` tag with some information about the commits. This information includes the `date` of the commit, the `description` if available, and within the tag there is a list of change tags. These changes consist of an `objectId` (e.g., `/test-project/test-project.capella#029c4531-1ece-4ea1-9b79-e61324b08faa`), which seems to be a combination of the project name + file + uuid and an optional `type` (e.g., `DELETED`, `MODIFIED`, `CREATED`, or non existing). Here is a small example of such an `activities` tag:

```xml
<?xml version="1.0" encoding="ASCII"?>
<activitymetadata:ActivityMetadataExport xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:activitymetadata="http://www.eclipse.org/sirius/activitymetadata/1.0.0" from="2024-01-23T15:31:36.293+0000" to="2024-01-23T15:44:40.265+0000">
  ...
  <activities description="[Export][Wizard]&#xA;" date="2024-01-23T15:32:44.113+0000">
    <changes objectId="/test-project/test-project.capella#029c4531-1ece-4ea1-9b79-e61324b08faa" type="MODIFIED"/>
    <changes objectId="/test-project/test-project.capella#f737664f-666d-43c3-8bad-42a7c8b04d3b" type="DELETED"/>
  </activities>
  ...
</activitymetadata:ActivityMetadataExport>
```

The second file is the `CommitHistory_[...].json` file, which contains essentially the same information in json format but is a bit larger due to the json overhead.

The main problem with including this commit information is that the CLI import can take a very long time and eventually times out. We have observed situations where these imports could take between 3 and 6 hours, which is not practical in our scenario. However, there seems to be a few ways we can make this work depending on the specific requirements.

#### Case 1 ####

The first case considers the current scenario, in which we want to have the entire detailed commit history in the Git repository. In this case, we can try to use the CLI arguments `-from` and `-to` of the importer. The `-from` argument specifies the starting time from which the metadata (i.e., commit history) should be imported, and the `-to` argument specifies the timestamp of the last commit used to export the metadata (if omitted, the last commit is used). The documentation ([4.3. Importer Configuration](https://www.obeosoft.com/download/release/team-for-capella/6.0/6.0.0/documentation/)) also states that the `-from` "can also be computed from an activity metadata model", in which case the parameter must be a URI or a path in the file system.

To realize this, we could first clone the Git repository and then either

- get the timestamp of the last commit from the existing commit history files (if any) and use it for the `-from` argument
- see if we can simply pass the path of the `.activitymetadata` file (if it already exists) to the `-from` argument to see if it automatically takes the last commit/timestamp from the file

Both approaches have problems with models that already have an extensive commit history but have never been pushed to Git. In this case, we can either try to import the entire existing commit history with the first run or omit the existing history.
We would also need to somehow merge/combine the commit history files on Git and the newly imported files.

#### Case 2 ####

The second case considers the scenario where we are only interested in the (new) commit messages (i.e., descriptions). In this case, it should be sufficient to set the CLI arguments `-importCommitHistoryAsJson` and `-includeCommitHistoryChanges` to `false`, which will prevent the detailed commit history from being imported. But even if both arguments are set to false, after a successful import we get a file `CommitHistory_[...].activitymetadata` that only contains some high-level information (that is because of the `-exportCommitHistory` argument being `true` by default). More specifically, this file has the same structure as the example above, but without all the `changes` tags. Since we already create this file with our current backup script, it is obviously quite fast and practical to use even for larger models.
This file could then either be pushed to Git, but the more important use case here would be to use the `description` values to somehow create meaningful commit messages. Here the two biggest challenges would be the following

1. extract the descriptions from the file and put them into a good structure that can be used as a commit message
2. ensure that only descriptions that have not yet been used / pushed to Git are included (i.e., avoid duplicates). For this we could either decide to always push the file `[...].activitymetadata` and extract the timestamp of the last considered commit from it or determine the timestamp of the last successful backup job and use that. 


#### Back to the issue description ####

To me, the description sounds a bit like creating a separate commit for each T4C commit. I think this is quite hard to achieve, as there is no way to tell CLI that I only want to get the model up to a certain timestamp (i.e., the `-from` and `-to` arguments are only used for the metadata). Therefore, we can only create a single git commit with all the changes from the last commit up to that point and include the T4C commit descriptions in some way. 
For me, your description fits more with case 2, i.e., including all T4C commit descriptions from the last to the latest commit in the Git commit message. It would be great to hear your opinion on this and whether case 2 would be sufficient for you.

**Comment by MoritzWeber0** [comment]
> To me, the description sounds a bit like creating a separate commit for each T4C commit. I think this is quite hard to achieve, as there is no way to tell CLI that I only want to get the model up to a certain timestamp (i.e., the -from and -to arguments are only used for the metadata).

OBEO announced that this might improve in TeamForCapella 7.0.0, which could provide an endpoint to get all commits. In combination with the `-form` and `-to` arguments, this could be used to translate the T4C commits into Git commits.

 > For me, your description fits more with case 2, i.e., including all T4C commit descriptions from the last to the latest commit in the Git commit message. 

Case 2 is easier to implement and can already fulfill some short-term feature requests. Therefore, we should investigate case 2 first. Once case 2 is implemented and Capella 7.0.0 is released, we can look into case 1 too.

**Comment by dominik003** [comment]
Current update to the approach @MoritzWeber0 and I decided on and steps to achieve it:

- In case of the *local* file handler, we do not want and need specific commit information and therefore do not want to set the `-from` argument
- In case of the *git* file handler
  - The git repository shall be cloned to the correct location `/tmp/git` 
  - The latest commit time shall be extracted. For this, we consider any commit on the specified branch, i.e., `$GIT_REPO_BRANCH`, that contains `Backup` in its message.
  - The importer CLI shall be executed with the correct `-from` timestamp set, either
    - Directly take the last commit timestamp in case that it is considered exclusively
    - Add a small offset (~1ms) to the timestamp to not include and already used T4C commit message
  - In case the importer CLI provides an `java.lang.IllegalArgumentException: The first timestamp must be lower than or equal to the second timestamp.` exception, we can be rather certain that there was no new T4C commit. Therefore, we shall provide a meaningful log message and stop the backup script
  - Otherwise, the imported model shall be handled as before (i.e., unzipped and copied to git repository directory)
  - The commit messages shall be extracted from the `.activitymetadata` file, located inside the `/tmp/model` directory. For this, we can consider the `description` and optionally the `date` of the T4C commits
  - The commit shall be pushed to git using the most recent T4C commit date as timestamp (i.e., the git commit timestamp and the last T4C commit timestamp shall be equal)
